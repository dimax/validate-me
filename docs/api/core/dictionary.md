---
sidebarDepth: 0
---

# Dictionary

It loads the message associated to rules. It also can have messages for nonexistent rules (e.g. rules that run server-side-only but need an informative message).

This module provides a function to configure it once for the entire app in the main file.

## How it works

1. If the user has setted the `clientHandler` prop through the `setConfig` function, it invokes it.
2. If the client code doesn't have that rule, it fallbacks to the default dictionary folder.
3. If the rule still doesn't exists, it fallbacks to the `_unknownRule`. Which:
   1. First tries to get it through `clientHandler` prop.
   2. Or it fallbacks to the `_unknownRule` from the library.
4. It caches the message function.

## Exports

| Name          |   Type   | Used by end user |
| ------------- | :------: | :--------------: |
| `setConfig`   | Function |       Yes        |
| `getWarning`  | Function |        No        |
| `getMessage`  | Function |        No        |
| `loadMessage` | Function |        No        |

### setConfig

- Firm:

```ts
setConfig(config: {
  lang?: string = 'en',
  handler: (lang: string, name: string) => Promise<Module>,
}): void
```

The `handler` prop must import the dictionary file for the specific `rule` in the specific `lang`. Each of the imported files must export by default a function with the following firm:

```ts
ruleName(value, ...args): String
```

#### Example

```js
const len = (value, min, max) =>
  `The value has length "${
    String(value).length
  }" but needs to be between [${min}, ${max}]`;

export default len;
```

### getWarning

- Firm:

```ts
getWarning(
  rule: { name: string, args: Array<string> },
  value: string,
): string
```

Used by every packages. It receives an unknown rule's name and args, and returns an informative message.

### getMessage

- Firm:

```ts
getMessage(
  rule: { name: string, args: Array<string> },
  value: string,
): string
```

Used by every packages. It receives a known rule's name and args, and returns an informative message.

### loadMessage

- Firm:

```ts
loadMessage(name: string): Promise<void>
```

Used by every packages. It receives the name of the rule's message to be loaded. It loads it and caches it.
