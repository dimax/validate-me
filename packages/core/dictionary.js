import { onModNotFound } from './rules';

// Dev/Test only
let toCall = true;

/**
 * @type {ClientRuleHandler} client handler
 * @return {ModulePromise} Returns a validator
 */
let clientHandler = () => {
  if (process.env.NODE_ENV !== 'production' && toCall) {
    // eslint-disable-next-line no-console
    console.warn(
      "[dev-only] @validate-me: Client's dictionary handler not found.",
    );

    toCall = false;
  }

  return Promise.reject();
};

let lang = 'en';
const dictionary = { en: {} };
const extras = {};
/**
 * @type {{[key: string]: {[key: string}: boolean}}
 */
const loadingCache = { en: {} };

/**
 * @param {Object} config The new configuration.
 * @param {string} [config.lang] The new language.
 * @param {Function} [config.handler] The client's handler.
 *
 * @return {void}
 */
export function setConfig(config) {
  if (config.lang) {
    lang = config.lang;

    if (!dictionary[lang]) {
      dictionary[lang] = {};
      loadingCache[lang] = {};
    }
  }
  if (config.handler) {
    clientHandler = config.handler;
  }
  if (!extras[lang]) {
    loadExtras();
  }
}

export function getWarning({ name, args }, value) {
  const unknownRule = extras[lang].unknownRule;
  const fn = dictionary[lang][name];
  const warning = extras[lang].preWarning || '';

  if (!(fn || unknownRule)) {
    // eslint-disable-next-line no-console
    return console.error(
      `validate-me: Cannot render a warning for the unknown rule "${name}".` +
        ' The _extras file must contain the "unknownRule" prop.',
    );
  }

  return warning + (fn ? fn(value, ...args) : unknownRule(name, value));
}

export function getMessage({ name, args }, value) {
  const fn = dictionary[lang][name];

  return fn ? fn(value, ...args) : '';
}

export function loadMessage(name) {
  if (dictionary[lang][name] || loadingCache[lang][name]) {
    return Promise.resolve();
  }
  if (!extras[lang]) {
    loadExtras();
  }
  const cache = loadingCache[lang];

  cache[name] = true;

  return clientHandler(lang, name)
    .catch(onModNotFound(() => import(`./dictionaries/${lang}/${name}.js`)))
    .catch(
      onModNotFound(() => {
        throw new Error(`Unknown dictionary for the "${name}" rule.`);
      }),
    )
    .then(({ default: rule }) => {
      dictionary[lang][name] = rule;
    })
    .finally(() => {
      cache[name] = false;
    });
}

function loadExtras() {
  extras[lang] = {};

  clientHandler(lang, '_extras')
    .catch(onModNotFound(() => import(`./dictionaries/${lang}/_extras.js`)))
    .catch(onModNotFound(() => {}))
    .then(mod => {
      extras[lang] = mod.default;
    });
}
