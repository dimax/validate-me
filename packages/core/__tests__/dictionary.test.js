import { setConfig, loadMessage, getMessage, getWarning } from '../dictionary';

describe('core/dictionary', () => {
  const required = 'required';
  const unknown = 'unknown';
  const unknownRule = { name: unknown, args: [] };
  const reqRule = { name: required, args: [] };

  test('Loads an existing message twice in a row (cache must prevent two requests)', () => {
    const promise1 = loadMessage(required);
    const promise2 = loadMessage(required);

    return Promise.all([promise1, promise2]).then(() => {
      expect(getMessage(reqRule)).toMatchSnapshot();
    });
  });

  test('Tries to load an unknown message but gets an error', () =>
    loadMessage(unknown)
      .then(() => {
        throw new Error('Expect to NOT load a known rule!');
      })
      .catch(err => {
        expect(err).toMatchSnapshot();
      }));

  test('Renders a message as a warning and as an error', () => {
    expect(getWarning(reqRule)).toMatchSnapshot();
  });

  test("Returns undefined when an language doesn't have configured the `unknownRule` function", () => {
    setConfig({ lang: 'es', handler: () => Promise.reject() });

    expect(getWarning(unknownRule)).toMatchSnapshot();
  });

  test('Renders an unknown rule as a warning and error messages', () => {
    setConfig({ lang: 'en' });

    expect(getWarning(unknownRule)).toMatchSnapshot();
    expect(getMessage(unknownRule)).toMatchSnapshot();
  });
});
