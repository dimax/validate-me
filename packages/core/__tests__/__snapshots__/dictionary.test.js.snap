// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`core/dictionary Loads an existing message twice in a row (cache must prevent two requests) 1`] = `"The field is required."`;

exports[`core/dictionary Renders a message as a warning and as an error 1`] = `"WARNING: The field is required."`;

exports[`core/dictionary Renders an unknown rule as a warning and error messages 1`] = `"WARNING: Unknown server rule \\"unknown\\" failed with the value \\"undefined\\""`;

exports[`core/dictionary Renders an unknown rule as a warning and error messages 2`] = `""`;

exports[`core/dictionary Returns undefined when an language doesn't have configured the \`unknownRule\` function 1`] = `undefined`;

exports[`core/dictionary Tries to load an unknown message but gets an error 1`] = `[Error: Unknown dictionary for the "unknown" rule.]`;
